cmd/milmove/main.go

```
@@ -826,6 +826,101 @@ func serveFunction(cmd *cobra.Command, args []string) error {
 	return nil
 }
 
+// initMigrateFlags - Order matters!
+func initMigrateFlags(flag *pflag.FlagSet) {
+	// Migration Config
+	cli.InitMigrationFlags(flag)
+
+	// DB Config
+	cli.InitDatabaseFlags(flag)
+
+	// Verbose
+	cli.InitVerboseFlags(flag)
+
+	// Don't sort flags
+	flag.SortFlags = false
+}
+
+func checkMigrateConfig(v *viper.Viper, logger logger) error {
+
+	logger.Info("checking migration config")
+
+	if err := cli.CheckMigration(v); err != nil {
+		return err
+	}
+
+	if err := cli.CheckDatabase(v, logger); err != nil {
+		return err
+	}
+
+	if err := cli.CheckVerbose(v); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func migrateFunction(cmd *cobra.Command, args []string) error {
+
+	err := cmd.ParseFlags(os.Args[1:])
+	if err != nil {
+		return errors.Wrap(err, "Could not parse flags")
+	}
+
+	flag := cmd.Flags()
+
+	v := viper.New()
+	err = v.BindPFlags(flag)
+	if err != nil {
+		return errors.Wrap(err, "Could not bind flags")
+	}
+	v.SetEnvKeyReplacer(strings.NewReplacer("-", "_"))
+	v.AutomaticEnv()
+
+	dbEnv := v.GetString(cli.DbEnvFlag)
+
+	logger, err := logging.Config(dbEnv, v.GetBool(cli.VerboseFlag))
+	if err != nil {
+		log.Fatalf("Failed to initialize Zap logging due to %v", err)
+	}
+
+	fields := make([]zap.Field, 0)
+	if len(gitBranch) > 0 {
+		fields = append(fields, zap.String("git_branch", gitBranch))
+	}
+	if len(gitCommit) > 0 {
+		fields = append(fields, zap.String("git_commit", gitCommit))
+	}
+	logger = logger.With(fields...)
+	zap.ReplaceGlobals(logger)
+
+	logger.Info("migrator starting up")
+
+	err = checkMigrateConfig(v, logger)
+	if err != nil {
+		logger.Fatal("invalid configuration", zap.Error(err))
+	}
+
+	// Create a connection to the DB
+	dbConnection, err := cli.InitDatabase(v, logger)
+	if err != nil {
```

Should probably fail hard here, right?

---

Good point. In this case we don't want to continue if the DB is not working.

---

I think the logic was sound but I changed `logger.Warn` to `logger.Fatal` and updated the wording around these failures. I hope its more clear.